File: csharp/Hachi64.Tests/Hachi64Tests.cs
using System;
using System.Linq;
using System.Text;
using Xunit;

namespace Hachi64.Tests
{
    public class Hachi64Tests
    {
        [Fact]
        public void TestEncodeDecodeConsistency()
        {
            var testCases = new[]
            {
                "Hello, World!",
                "rust",
                "a",
                "ab",
                "abc",
                "Python",
                "Hachi64",
                "",
                "The quick brown fox jumps over the lazy dog"
            };

            foreach (var testData in testCases)
            {
                var bytes = Encoding.UTF8.GetBytes(testData);
                var encoded = Hachi64.Encode(bytes);
                var decoded = Hachi64.Decode(encoded);
                
                Assert.Equal(bytes, decoded);
            }
        }

        [Theory]
        [InlineData("Hello", "豆米啊拢嘎米多=")]
        [InlineData("abc", "西阿南呀")]
        [InlineData("a", "西律==")]
        [InlineData("ab", "西阿迷=")]
        [InlineData("Python", "抖咪酷丁息米都慢")]
        [InlineData("Hello, World!", "豆米啊拢嘎米多拢迷集伽漫咖苦播库迷律==")]
        [InlineData("Base64", "律苦集叮希斗西丁")]
        [InlineData("Hachi64", "豆米集呀息米库咚背哈==")]
        public void TestSpecificEncodings(string input, string expected)
        {
            var bytes = Encoding.UTF8.GetBytes(input);
            var encoded = Hachi64.Encode(bytes);
            
            Assert.Equal(expected, encoded);
        }

        [Fact]
        public void TestBinaryData()
        {
            var testData = Enumerable.Range(0, 256).Select(i => (byte)i).ToArray();
            var encoded = Hachi64.Encode(testData);
            var decoded = Hachi64.Decode(encoded);
            
            Assert.Equal(testData, decoded);
        }

        [Fact]
        public void TestNoPaddingEncodeDecod()
        {
            var testCases = new[] { "a", "ab", "abc", "Hello" };

            foreach (var testData in testCases)
            {
                var bytes = Encoding.UTF8.GetBytes(testData);
                var encodedNoPad = Hachi64.Encode(bytes, padding: false);
                var decoded = Hachi64.Decode(encodedNoPad, padding: false);
                
                Assert.Equal(bytes, decoded);
            }
        }

        [Fact]
        public void TestPaddingBehavior()
        {
            var testData = Encoding.UTF8.GetBytes("a");

            var encodedWithPadding = Hachi64.Encode(testData, padding: true);
            var encodedWithoutPadding = Hachi64.Encode(testData, padding: false);

            // 有填充的应该有 == 结尾
            Assert.EndsWith("==", encodedWithPadding);

            // 无填充的不应该有 = 
            Assert.DoesNotContain("=", encodedWithoutPadding);

            // 两种方式解码应该都能得到原始数据
            var decodedWithPadding = Hachi64.Decode(encodedWithPadding, padding: true);
            var decodedWithoutPadding = Hachi64.Decode(encodedWithoutPadding, padding: false);

            Assert.Equal(testData, decodedWithPadding);
            Assert.Equal(testData, decodedWithoutPadding);
        }

        [Fact]
        public void TestDecodeInvalidInput()
        {
            // 测试包含不在字母表中的字符
            Assert.Throws<ArgumentException>(() => Hachi64.Decode("包含无效字符X的字符串"));

            // 测试空字符串
            var result = Hachi64.Decode("");
            Assert.Empty(result);
        }

        [Fact]
        public void TestAlphabetCoverage()
        {
            // 验证字母表长度
            Assert.Equal(64, Hachi64.HachiAlphabet.Length);

            // 验证字母表无重复字符
            var uniqueChars = Hachi64.HachiAlphabet.Distinct().Count();
            Assert.Equal(64, uniqueChars);

            // 测试长数据以确保所有字符都可能被使用
            var longData = Enumerable.Range(0, 256).SelectMany(_ => Enumerable.Range(0, 256).Select(i => (byte)i)).ToArray();
            var encoded = Hachi64.Encode(longData);
            var decoded = Hachi64.Decode(encoded);
            
            Assert.Equal(longData, decoded);
        }

        [Fact]
        public void TestEmptyInput()
        {
            var emptyBytes = Array.Empty<byte>();
            var encoded = Hachi64.Encode(emptyBytes);
            var decoded = Hachi64.Decode(encoded);
            
            Assert.Empty(encoded);
            Assert.Empty(decoded);
        }

        [Fact]
        public void TestNullInput()
        {
            Assert.Throws<ArgumentNullException>(() => Hachi64.Encode(null!));
            Assert.Throws<ArgumentNullException>(() => Hachi64.Decode(null!));
        }

        [Theory]
        [InlineData(1)]
        [InlineData(2)]
        [InlineData(3)]
        [InlineData(10)]
        [InlineData(100)]
        [InlineData(1000)]
        public void TestVariousLengths(int length)
        {
            var testData = Enumerable.Range(0, length).Select(i => (byte)(i % 256)).ToArray();
            var encoded = Hachi64.Encode(testData);
            var decoded = Hachi64.Decode(encoded);
            
            Assert.Equal(testData, decoded);
        }

        [Fact]
        public void TestRoundtripWithLongText()
        {
            var testData = Encoding.UTF8.GetBytes("The quick brown fox jumps over the lazy dog. 这是一段中文文本测试。1234567890!@#$%^&*()");
            var encoded = Hachi64.Encode(testData);
            var decoded = Hachi64.Decode(encoded);
            var decodedText = Encoding.UTF8.GetString(decoded);
            
            Assert.Equal(testData, decoded);
            Assert.Equal("The quick brown fox jumps over the lazy dog. 这是一段中文文本测试。1234567890!@#$%^&*()", decodedText);
        }
    }
}

File: swift/Tests/Hachi64Tests/Hachi64Tests.swift
import XCTest
@testable import Hachi64

final class Hachi64Tests: XCTestCase {
    
    // MARK: - Test Encode Examples from Documentation
    
    func testEncodeHachi64Examples() {
        let testCases: [(String, String)] = [
            ("Hello", "豆米啊拢嘎米多="),
            ("abc", "西阿南呀"),
            ("Python", "抖咪酷丁息米都慢"),
            ("Hello, World!", "豆米啊拢嘎米多拢迷集伽漫咖苦播库迷律=="),
            ("Base64", "律苦集叮希斗西丁"),
            ("Hachi64", "豆米集呀息米库咚背哈==")
        ]
        
        for (input, expected) in testCases {
            let data = input.data(using: .utf8)!
            let result = Hachi64.encode(data)
            XCTAssertEqual(result, expected, "Encode(\(input)) = \(result), want \(expected)")
        }
    }
    
    // MARK: - Test Decode Examples from Documentation
    
    func testDecodeHachi64Examples() {
        let testCases: [(String, String)] = [
            ("豆米啊拢嘎米多=", "Hello"),
            ("西阿南呀", "abc"),
            ("抖咪酷丁息米都慢", "Python"),
            ("豆米啊拢嘎米多拢迷集伽漫咖苦播库迷律==", "Hello, World!"),
            ("律苦集叮希斗西丁", "Base64"),
            ("豆米集呀息米库咚背哈==", "Hachi64")
        ]
        
        for (input, expected) in testCases {
            do {
                let result = try Hachi64.decode(input)
                let resultString = String(data: result, encoding: .utf8)!
                XCTAssertEqual(resultString, expected, "Decode(\(input)) = \(resultString), want \(expected)")
            } catch {
                XCTFail("Decode(\(input)) returned error: \(error)")
            }
        }
    }
    
    // MARK: - Test Edge Cases for Encoding
    
    func testEncodeEdgeCases() {
        let testCases: [(Data, String, String)] = [
            (Data(), "", "empty"),
            ("a".data(using: .utf8)!, "西律==", "single byte"),
            ("ab".data(using: .utf8)!, "西阿迷=", "two bytes"),
            ("abc".data(using: .utf8)!, "西阿南呀", "three bytes")
        ]
        
        for (input, expected, name) in testCases {
            let result = Hachi64.encode(input)
            XCTAssertEqual(result, expected, "\(name): Encode(\(String(data: input, encoding: .utf8) ?? "")) = \(result), want \(expected)")
        }
    }
    
    // MARK: - Test Edge Cases for Decoding
    
    func testDecodeEdgeCases() {
        let testCases: [(String, Data, String)] = [
            ("", Data(), "empty"),
            ("西律==", "a".data(using: .utf8)!, "single byte"),
            ("西阿迷=", "ab".data(using: .utf8)!, "two bytes"),
            ("西阿南呀", "abc".data(using: .utf8)!, "three bytes")
        ]
        
        for (input, expected, name) in testCases {
            do {
                let result = try Hachi64.decode(input)
                XCTAssertEqual(result, expected, "\(name): Decode(\(input)) mismatch")
            } catch {
                XCTFail("\(name): Decode(\(input)) returned error: \(error)")
            }
        }
    }
    
    // MARK: - Test Invalid Input
    
    func testDecodeInvalidInput() {
        let testCases = [
            "ABC",
            "哈哈哈X",
            "123"
        ]
        
        for input in testCases {
            do {
                _ = try Hachi64.decode(input)
                XCTFail("Decode(\(input)) should return error for invalid input")
            } catch {
                // Expected to throw
                XCTAssertTrue(error is Hachi64Error, "Expected Hachi64Error, got \(error)")
            }
        }
    }
    
    // MARK: - Test Roundtrip
    
    func testRoundtrip() {
        let testCases = [
            "Hello, World!",
            "rust",
            "a",
            "ab",
            "abc",
            "Python",
            "Hachi64",
            "",
            "The quick brown fox jumps over the lazy dog"
        ]
        
        for input in testCases {
            let data = input.data(using: .utf8)!
            let encoded = Hachi64.encode(data)
            
            do {
                let decoded = try Hachi64.decode(encoded)
                XCTAssertEqual(decoded, data, "Roundtrip failed for: \(input)")
            } catch {
                XCTFail("Decode returned error for \(input): \(error)")
            }
        }
    }
    
    // MARK: - Test Binary Data
    
    func testBinaryData() {
        // Create binary data with all byte values 0-255
        var binaryData = Data()
        for i in 0..<256 {
            binaryData.append(UInt8(i))
        }
        
        let encoded = Hachi64.encode(binaryData)
        
        do {
            let decoded = try Hachi64.decode(encoded)
            XCTAssertEqual(decoded, binaryData, "Binary data roundtrip failed")
        } catch {
            XCTFail("Decode returned error: \(error)")
        }
    }
    
    // MARK: - Test No Padding
    
    func testNoPaddingEncodeDecode() {
        let testCases = [
            "a",
            "ab",
            "abc",
            "Hello"
        ]
        
        for input in testCases {
            let data = input.data(using: .utf8)!
            let encodedNoPad = Hachi64.encode(data, padding: false)
            
            do {
                let decoded = try Hachi64.decode(encodedNoPad, padding: false)
                XCTAssertEqual(decoded, data, "No-padding roundtrip failed for: \(input)")
            } catch {
                XCTFail("Decode returned error for \(input): \(error)")
            }
        }
    }
    
    // MARK: - Test Padding Behavior
    
    func testPaddingBehavior() {
        let testData = "a".data(using: .utf8)!
        
        let encodedWithPadding = Hachi64.encode(testData, padding: true)
        let encodedWithoutPadding = Hachi64.encode(testData, padding: false)
        
        // With padding should end with ==
        XCTAssertTrue(encodedWithPadding.hasSuffix("=="), "Encoded with padding should end with ==, got \(encodedWithPadding)")
        
        // Without padding should not contain =
        XCTAssertFalse(encodedWithoutPadding.contains("="), "Encoded without padding should not contain =, got \(encodedWithoutPadding)")
        
        // Both should decode to the same result
        do {
            let decodedWithPadding = try Hachi64.decode(encodedWithPadding, padding: true)
            let decodedWithoutPadding = try Hachi64.decode(encodedWithoutPadding, padding: false)
            
            XCTAssertEqual(decodedWithPadding, testData, "Decoded with padding should equal original data")
            XCTAssertEqual(decodedWithoutPadding, testData, "Decoded without padding should equal original data")
        } catch {
            XCTFail("Decode returned error: \(error)")
        }
    }
    
    // MARK: - Test Alphabet Coverage
    
    func testAlphabetCoverage() {
        // Verify alphabet length
        let alphabetArray = Array(Hachi64.hachiAlphabet)
        XCTAssertEqual(alphabetArray.count, 64, "Alphabet length should be 64")
        
        // Verify no duplicate characters
        let uniqueChars = Set(alphabetArray)
        XCTAssertEqual(uniqueChars.count, 64, "Alphabet should contain 64 unique characters")
        
        // Test long data to ensure all characters can be used
        var longData = Data()
        for _ in 0..<3 {
            for i in 0..<256 {
                longData.append(UInt8(i))
            }
        }
        
        let encoded = Hachi64.encode(longData)
        
        do {
            let decoded = try Hachi64.decode(encoded)
            XCTAssertEqual(decoded, longData, "Long data roundtrip failed")
        } catch {
            XCTFail("Decode returned error: \(error)")
        }
    }
}
